cachito/workers/pkg_managers/pip.py

```diff
@@ -2175,3 +2185,67 @@ def get_raw_component_name(requirement):
         raw_component_name = f"{repo_name}/{filename}"
 
     return raw_component_name
+
+
+def iter_zip_file(filename):
+    with zipfile.ZipFile(filename, "r") as zf:
+        return zf.namelist()
+
+
+def iter_tar_z(filename):
+    # uncompress file > another_file
+    with tempfile.TemporaryDirectory() as work_dir:
+        test_file = os.path.join(work_dir, os.path.basename(filename))
+        with open(test_file, "wb") as f:
+            subprocess.check_call(["uncompress", "-c", filename], stdout=f)
+        return iter_tar_file(test_file)
+
+
+def iter_tar_file(filename):
+    with tarfile.open(filename, "r") as tar:
+        for member in tar:
+            yield member.name
+
+
+def is_pkg_info_dir(path):
+    """Simply check whether a path represents the PKG_INFO directory
+
+    Generally, it is in the format for example: pkg-1.0/PKG_INFO
+
+    :param str path: a path.
+    :return: True if it is, otherwise False is returned.
+    :rtype: bool
+    """
+    parts = os.path.split(path)
+    return len(parts) == 2 and parts[1] == "PKG_INFO"
+
+
+def check_metadata_in_sdist(sdist_path):
+    """Check if a sdist package has metadata
+
+    :param str sdist_path: the path of a sdist package file.
+    :return: True if has, otherwise False is returned.
+    :rtype: bool
+    :raise ValidationError: if the sdist cannot be validated.
+    """
+    try:
+        if sdist_path.endswith(ZIP_FILE_EXT):
+            files_iter = iter_zip_file(sdist_path)
+        elif sdist_path.endswith(COMPRESSED_TAR_EXT):
+            files_iter = iter_tar_z(sdist_path)
+        elif any(map(sdist_path.endswith, SDIST_FILE_EXTENSIONS)):
+            files_iter = iter_tar_file(sdist_path)
+        else:
+            raise ValidationError(
+                f"Cannot check metadata from {sdist_path}, "
+                f"which does not have a known supported extension.",
+            )
+    except tarfile.ReadError as e:
+        raise ValidationError(f"Cannot open {sdist_path} as a Tar file. Error: {str(e)}")
+    except zipfile.BadZipFile:
+        raise ValidationError(f"Cannot open {sdist_path} as a Zip file. Error: {str(e)}")
+    except subprocess.CalledProcessError as e:
+        raise ValidationError(f"Cannot uncompress {sdist_path} to validate. Error: {str(e)}")
+    if not any(map(is_pkg_info_dir, files_iter)):
+        filename = os.path.basename(sdist_path)
+        raise ValidationError(f"{filename} does not include metadata.")
```

Comment:
> It would be awesome if the error message could include the suggested download url.

Nevermind, I forgot that the Cachito download url points to Nexus, not to PyPI. Telling users to get the package from git or a direct download url will have to be good enough.

Reply:
Updated the message. But, not sure if the update is accurate or informative enough. PTAL.

Reply:
I would go with something like this:

{sdist_path.name} does not include metadata (there is no PKG-INFO file). It is not a valid sdist and cannot be downloaded from PyPI. Consider editing your requirements file to download the package from git or a direct download URL instead.

Reply:
It's clear. Just update it to the code. Thanks.